### PR TITLE
fix: Safe snowflake and dynamic view col types

### PR DESCRIPTION
**Pull request by ianedwards**
## What is the problem?

- Snowflake was not using safely escaped parameters in queries
- View columns were not created with the correct data type from the source

## Changes being made

- Created a SnowflakeClient which promisifies the snowflake-sdk and enables a connection to be created that can match the return values for connections.ts
- Updated all Snowflake queries to be wrapped in sql template tag

## How did you test this code?

- Manual test

### CODE REVIEW

File: lib/connections.ts [hunk 1]
```diff
@@ -50,86 +57,60 @@ export const getConnection = async ({
...
-          message: "Database type is not currently supported.",
+          code: "not_implemented",
+          message: `Database type ${dbType} has not yet been implemented.`,
         };
       }
     }
-  } catch (err) {
-    const message = err instanceof Error ? err.message : String(err);
+  } catch (error) {
+    logger.info({ connectionError: error });
```
Comment (timothygoltser) [inline lib/connections.ts:104]: Use logger.error

File: server/queues/transfer/processor.ts [hunk 2]
```diff
@@ -215,6 +215,7 @@ export default async function (job: Job<TransferQueueJobData>) {
       case "SNOWFLAKE": {
         const credentialsExist =
           !!destHost &&
+          !!destPort &&
```
Comment (timothygoltser) [inline server/queues/transfer/processor.ts:218]: Boolean(... && ...)

File: server/queues/transfer/processor.ts [hunk 3]
```diff
@@ -251,18 +254,19 @@ export default async function (job: Job<TransferQueueJobData>) {
...
           credentials = (aws_key_id='${
             env.S3_USER_ACCESS_ID
           }' aws_secret_key='${env.S3_USER_SECRET_KEY}')
-          encryption = (TYPE='AWS_SSE_KMS' KEY='${env.KMS_KEY_ID}')
+          encryption = (TYPE='AWS_SSE_KMS' KMS_KEY_ID='${env.KMS_KEY_ID}')
           file_format = (TYPE='CSV' FIELD_DELIMITER=',' SKIP_HEADER=1);
-        `;
-        await destConnection.client.query(createStageOperation);
+        `);
+        await destConnection.query(createStageOperation);
```
Comment (timothygoltser) [inline server/queues/transfer/processor.ts:269]: Could be a queryUnsafe

File: lib/snowflake/client.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+export class CustomSnowflakeError extends Error {
+  constructor(base: SnowflakeError) {
+    super(base.message);
+  }
+}
+
+class SnowflakeClient {
+  private connection;
```
Comment (timothygoltser) [inline lib/snowflake/client.ts:24]: Use #connection for parity with JS standard

File: server/routes/views/index.ts [hunk 5]
```diff
@@ -122,49 +122,61 @@ viewRouter.post("/", async (req, res: ApiResponse<ViewResponse>) => {
   const { sourceType, host, port, schema, database, username, password } =
     source;
 
-  const test = await testQuery({
+  const test = await useConnection({
```
Comment (timothygoltser) [inline server/routes/views/index.ts:125]: This is a generic connection so something like 'conn' is easier to understand

**Review by timothygoltser — CHANGES_REQUESTED** [review]

File: server/queues/transfer/processor.ts (hunk 2 above)
Comment (ianedwards) [reply server/queues/transfer/processor.ts:218]: not friendly with TS compiler

**Review by timothygoltser — APPROVED** [review]

**Pull request merged**